### PR TITLE
Support emsdk --embedded mode

### PR DESCRIPTION
**Pull request by sbc100**
This change allows `emsdk --embdded` mode to work out of the box
without running emsdk_env.sh.

This paves makes it easier to justify using embedded mode by default:
  https://github.com/emscripten-core/emsdk/pull/472

This is part of a larger plan to avoid using the $HOME directory to
store anything at all: #9543

### CODE REVIEW

File: tools/shared.py [hunk 1]
```diff
@@ -265,6 +265,12 @@ def generate_config(path, first_time=False):
 # 4. Fall back users home directory (~/.emscripten).
 
 embedded_config = path_from_root('.emscripten')
+# For compatability with emsdk --embedded mode also look two levels up.  This
```
Comment (waywardmonkeys) [inline tools/shared.py:268]: Typo for 'compatibility'

File: tools/shared.py [hunk 2]
```diff
@@ -265,6 +265,12 @@ def generate_config(path, first_time=False):
...
+# could be removed in emsdk was use the emscripten directory itself for the
```
Comment (kripken) [inline tools/shared.py:269]: I don't follow this statement - should `in` be `if`, and `use` be `using`?

But I'm not sure I get it. When are `embedded_config` and `emsdk_embedded_config` different? Clarifying that would be useful here I think.

Comment (sbc100) [reply tools/shared.py]: Thanks, fixed the grammar there.

The issue is that emsdk's --embedded mode currently always puts the .emscripton config file in the root of the emsdk, which is two levels up from the emscripten directory.  I'm working on trying to change that but it seems a little complicated.

emscripten already has support for an embedded config, but only its own directory.

This change is only really needs for users who don't run `emsdk_env.sh` and just run `emcc` directly.

Most users are expected to run `emsdk_env.sh` to use the SDK, but it turns out the if you just run `/path/to/emscripten/ecc` without running `emsdk_env.sh` today it often just works because emscripten knowns how to find its config and everything flows from there.   Without this change if you use `emsdk --embedded` and then run `/path/to/emcc` *without* having first run `emsdk_env.sh` then emscripten will not find any config file at all and default back to its method of create new config file with default settings in $HOME.  

This change means that emsdk users won't see this default bahaviour becuse them will look in the correct plance for the embedded config file, even if EM_CONFIG is not set (which it normally is when you run `emsdk_env.sh`.

Sorry for the long explanation, its kind of annoyingly complicated.

**Comment by sbc100** [comment]
tldr; This change is the minimize the impact of switch to `--embedded` by default in the emsdk.

File: tools/shared.py [hunk 3]
```diff
@@ -265,6 +265,12 @@ def generate_config(path, first_time=False):
 # 4. Fall back users home directory (~/.emscripten).
 
 embedded_config = path_from_root('.emscripten')
+# For compatibility with emsdk --embedded mode also look two levels up.  This
```
Comment (kripken) [inline tools/shared.py:268]: maybe mention why two levels up? That is, that when using the emsdk, the structure would look like
```
emsdk/
  emscripten/
```
But that picture must be wrong since you say two levels and not one..?

**Comment by kripken** [comment]
The tl;dr sounds good! Sorry I don't get the technical details yet.

**Comment by sbc100** [comment]
Added more comments that should make it more obvious why we need this.

**Comment by kripken** [comment]
I'm sorry, I still don't get it :cry: why **2** levels up? Is it not this:

```
emsdk/
  emscripten/
```

Where is the extra directory? I really think it would be good to add a little drawing like that into the comment, it would have helped me get this a lot sooner.

**Comment by sbc100** [comment]
Also not that this change is more a convenience for emsdk users that forget to run `./emsdk_env.sh`.

emsdk users that correctly set thier environment will not be effected by this.

It also only effects users of `--embedded` which hopefully soon will be all users :)

**Comment by sbc100** [comment]
The places that emscripten lives with the emsdk look like this:

```
emsdk/
  fastcomp/
    emscripten/
```

```
emsdk/
  upstream/
    emscripten/
```

Or if building from source they look like:

```
emsdk/
  1.38.30/
    emscripten/
```

It never lives directly in `emsdk/emscripten`.

**Comment by sbc100** [comment]
I did add a lot of comments to the code already.  Did you look at the latest version?   Do you think it needs more?

File: tools/shared.py [hunk 4]
```diff
@@ -261,10 +261,24 @@ def generate_config(path, first_time=False):
 # The search order from the config file is as follows:
 # 1. Specified on the command line (--em-config)
 # 2. Specified via EM_CONFIG environment variable
-# 3. If emscripten-local .emscripten file is found, use that
-# 4. Fall back users home directory (~/.emscripten).
+# 3. Local .emscripten file, if found
+# 4. Local .emscripten file, as written by emsdk --embedded (two levels up), if found.
```
Comment (kripken) [inline tools/shared.py:265]: It says "two levels" here, but without a little graph like in that comment of yours, it's very unclear I think.

likewise, on line 269 it also says "two levels" but doesn't explain.

If I'm being unclear here, I apologize, and feel free to land this - I'll open a followup PR.

**Comment by sbc100** [comment]
I included some exampels.

**Comment by kripken** [comment]
Thanks!
